compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp

```diff
@@ -0,0 +1,212 @@
+        if (matchPattern(offset, m_Constant()))
+          continue;
+        else if (llvm::is_contained(indVars, offset))
+          continue;
+        // We assume that if a subview has an offset which is not
+        // a constant and not a thread id, it's not 'distributing'.
+        else
+          return {};
+      }
+
+      auto nextType = cast<MemRefType>(subview.getResult().getType());
+      if (!type) {
+        type = nextType;
+      } else if (type != nextType) {
+        // This is the case where there are 2+ subview ops which look like
+        // they should be distributing, but they have different result types.
+        // Bail.
+        return {};
+      }
+    }
+  }
+  return type;
+}
+
+/// Create a copy of `toUpdate` with all values in `toRemove` replaced by
+/// `replacement`.
+template <typename Container>
+SmallVector<Value> substitute(Container toUpdate,
+                              const DenseSet<Value> &toRemove,
+                              Value replacement) {
+  SmallVector<Value> updated(toUpdate.begin(), toUpdate.end());
+  for (Value &v : updated) {
+    if (toRemove.contains(v)) v = replacement;
+  }
+  return updated;
+}
+
+/// Distribute local memory accesses through subviews by allocating a single,
+/// smaller memory. This is ultimately needed because cores can't operate on
+/// one shared L1 memory.
+LogicalResult distributeLocalMemory(ModuleOp moduleOp) {
+  FailureOr<DenseSet<Value>> maybeIndVars = getThreadIndVars(moduleOp);
+  if (failed(maybeIndVars)) return failure();
+  const DenseSet<Value> &indVars = maybeIndVars.value();
+  IRRewriter rewriter(moduleOp.getContext());
+  moduleOp->walk([&](memref::AllocOp oldAlloc) {
+    // Only consider local memory (L1).
+    Attribute maybeMemorySpace = oldAlloc.getType().getMemorySpace();
+    if (!maybeMemorySpace) return WalkResult::advance();
+    auto memorySpace = cast<IntegerAttr>(maybeMemorySpace);
+    if (memorySpace.getInt() != 2) return WalkResult::advance();
+
+    // Don't try and distribute memory if the alloc is inside a scf.for op.
+    if (auto scfForOp = oldAlloc->getParentOfType<scf::ForOp>())
+      return WalkResult::advance();
+
+    MemRefType memRefType = getDistributedType(oldAlloc, indVars);
+
+    // Failed to find a memref.subview that looks like it is distributing.
+    // This doesn't mean that we can't distribute (for example there might be
+    // no subviews at all), but this requires further work.
+    if (!memRefType) return WalkResult::advance();
+
+    ArrayRef<int64_t> newShape = memRefType.getShape();
+    Type elementType = memRefType.getElementType();
+
+    rewriter.setInsertionPoint(oldAlloc);
+    MemRefType newAllocType = MemRefType::get(
+        newShape, elementType, MemRefLayoutAttrInterface{}, memorySpace);
+    auto newAlloc = rewriter.create<memref::AllocOp>(rewriter.getUnknownLoc(),
+                                                     newAllocType);
+
+    auto users_ = oldAlloc->getUsers();
```

Comment:
Trailing underscores are widely used for class member variables, so I think we should avoid using it.